File: go/detect_capital.go
package leetcode

// Time complexity: O(n) where n is the length of word
// Space complexity: O(1)
func detectCapitalUse(word string) bool {
	return allLower(word) || allUpper(word) || capitalizedWord(word)
}

func allLower(word string) bool {
	for _, el := range word {
		if el < 'a' || el > 'z' {
			return false
		}
	}
	return true
}

func allUpper(word string) bool {
	for _, el := range word {
		if el < 'A' || el > 'Z' {
			return false
		}
	}
	return true
}

func capitalizedWord(word string) bool {
	if word[0] < 'A' || word[0] > 'Z' {
		return false
	}

	for i := 1; i < len(word); i++ {
		if word[i] < 'a' || word[i] > 'z' {
			return false
		}
	}
	return true
}

File: python/detect-capital.py
class Solution:
    # Note about space complexity of all the below algorithms:
    # All of them are in fact O(n) because of word[1:] which makes
    # a copy of word dropping the first character
    # This can be avoided by just looping over those characters
    
    # Time complexity: O(n) where n is the length of word
    # Space complexity: O(1)
    def detectCapitalUse(self, word: str) -> bool:
        return (
            all(map(lambda c: ord('a') <= ord(c) <= ord('z'), word)) or
            all(map(lambda c: ord('A') <= ord(c) <= ord('Z'), word)) or
            (
                ord('A') <= ord(word[0]) <= ord('Z') and
                all(map(lambda c: ord('a') <= ord(c) <= ord('z'), word[1:]))
            )
        )
    # Time complexity: O(n) where n is the length of word
    # Space complexity: O(1)
    def detectCapitalUse2(self, word: str) -> bool:
        return word.islower() or word.isupper() or (word[0].isupper() and word[1:].islower())
    
    
    # Time complexity: O(n) where n is the length of word
    # Space complexity: O(1)
    def detectCapitalUse3(self, word: str) -> bool:
        return all(map(str.islower, word)) or all(map(str.isupper, word)) or (word[0].isupper() and word[1:].islower())